modules/tensor_mechanics/src/materials/ComputeThermalExpansionEigenstrainBase.C

```diff
@@ -10,41 +10,85 @@
 #include "ComputeThermalExpansionEigenstrainBase.h"
 #include "RankTwoTensor.h"
 
+template <bool is_ad>
 InputParameters
-ComputeThermalExpansionEigenstrainBase::validParams()
+ComputeThermalExpansionEigenstrainBaseTempl<is_ad>::validParams()
 {
-  InputParameters params = ComputeEigenstrainBase::validParams();
+  InputParameters params = ComputeEigenstrainBaseTempl<is_ad>::validParams();
   params.addCoupledVar("temperature", "Coupled temperature");
   params.addRequiredCoupledVar("stress_free_temperature",
                                "Reference temperature at which there is no "
                                "thermal expansion for thermal eigenstrain "
                                "calculation");
+  params.addParam<bool>("use_old_temperature",
+                        false,
+                        "Flag to optionally use the temperature value from the previous timestep.");
   return params;
 }
 
-ComputeThermalExpansionEigenstrainBase::ComputeThermalExpansionEigenstrainBase(
+template <bool is_ad>
+ComputeThermalExpansionEigenstrainBaseTempl<is_ad>::ComputeThermalExpansionEigenstrainBaseTempl(
     const InputParameters & parameters)
-  : DerivativeMaterialInterface<ComputeEigenstrainBase>(parameters),
-    _temperature(coupledValue("temperature")),
-    _deigenstrain_dT(declarePropertyDerivative<RankTwoTensor>(_eigenstrain_name,
-                                                              getVar("temperature", 0)->name())),
-    _stress_free_temperature(coupledValue("stress_free_temperature"))
+  : DerivativeMaterialInterface<ComputeEigenstrainBaseTempl<is_ad>>(parameters),
+    _use_old_temperature(this->template getParam<bool>("use_old_temperature")),
+    _temperature_old(_use_old_temperature ? this->coupledValueOld("temperature") : this->_zero),
+    _temperature(this->template coupledGenericValue<is_ad>("temperature")),
+    _deigenstrain_dT(std::invoke(
+        [&]() -> decltype(auto)
```

Comment:
🤩

Reply:
The decltype(auto) makes sure the return type is a reference in the non ad case. Without it the compiler deduces return by value, which fails for the material property.

Reply:
![image](https://user-images.githubusercontent.com/7014706/166606453-750eda84-68bf-4dad-b5a9-9a73ec2b90e0.png)
MAGIC!